## STA130 HW 1

Rasyid Rafi Pamuji

Student Number: 1011270081

1. Using the dataset from the STA130 Instructions page:

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

2. Using ChatGPT, I created a new session with the prompt of asking how many columns and rows of data a "pandas" Dataframe has. The prompt I used was "Hey ChatGPT, I have a dataset and I need help on how many columns and rows of data a "pandas" DataFrame has. This is the dataset:
"https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv". Specifiy which is which."

Code provided:

In [2]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
villagers_df = pd.read_csv(url)

# Get the number of rows and columns
rows, columns = villagers_df.shape
print(f"Rows: {rows}, Columns: {columns}")

Rows: 391, Columns: 11


From my own understanding of the general definitions, an "observation" is each unique datapoint being measured in each of the rows in the dataset. In my dataset, each row represents a unique "villager". The "variables" are each of the characteristics or features given to the "observations", in my case, the villagers, in each of the columns in the dataset.

3. When asking the Chatbot for summaries of the columns in the dataset using: df.describe() and df['column'].value_counts(), it creates these codes:

In [14]:
# General summary for numeric and categorical data
print(df.describe(include='all'))

# Count the number of villagers for each species
print(df['species'].value_counts())

# Count the number of villagers for each personality type
print(df['personality'].value_counts())

             row_n       id     name gender species birthday personality  \
count   391.000000      390      391    391     391      391         391   
unique         NaN      390      391      2      35      361           8   
top            NaN  admiral  Admiral   male     cat     1-27        lazy   
freq           NaN        1        1    204      23        2          60   
mean    239.902813      NaN      NaN    NaN     NaN      NaN         NaN   
std     140.702672      NaN      NaN    NaN     NaN      NaN         NaN   
min       2.000000      NaN      NaN    NaN     NaN      NaN         NaN   
25%     117.500000      NaN      NaN    NaN     NaN      NaN         NaN   
50%     240.000000      NaN      NaN    NaN     NaN      NaN         NaN   
75%     363.500000      NaN      NaN    NaN     NaN      NaN         NaN   
max     483.000000      NaN      NaN    NaN     NaN      NaN         NaN   

                song   phrase           full_id  \
count            380      391       

4. I can find out if my dataset has non-numeric variables and missing values in numeric variables by using this code below:

In [18]:
# Check the data types of all columns
print(df.dtypes)

# Check for missing values in numeric columns
print(df.select_dtypes(include=['number']).isnull().sum())

row_n           int64
id             object
name           object
gender         object
species        object
birthday       object
personality    object
song           object
phrase         object
full_id        object
url            object
dtype: object
row_n    0
dtype: int64


Non-numeric variables contain data types like "object". My dataset has no numeric variables with missing values as row_n is the only column and it has 0 missing values.

Instead, I will use the "titanic.csv" for further analysis into the topic.

In [ ]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
titanic_df = pd.read_csv(url)

# Get the shape of the dataset
shape = titanic_df.shape
print(f"Shape of the dataset (rows, columns): {shape}")

# Get the descriptive statistics for numeric columns
describe = titanic_df.describe()
print("\nDescriptive statistics for numeric columns:")
print(describe)

# Check the count column from df.describe()
print("\n'Count' of non-missing values for each numeric column:")
print(describe.loc['count'])

Based on the coding analysis:
- df.shape gives the total number of rows and columns.
- df.describe() provides statistics for numeric columns only.
- describe.loc['count'] shows the number of non-missing values for each numeric column.

- df.shape shows 15 columns, but df.describe() only analyzes the numeric columns (6 in this case).
- The age column has 714 non-missing values, as shown by the count in df.describe(), meaning 177 values are missing (891 - 714 = 177).

Therefore, the discrepancies explained are:
- df.shape reports the total number of rows and columns, including all variables (numeric and non-numeric).
- df.describe() reports only numeric columns and gives a count of non-missing values, so it might show fewer columns and a smaller count of rows due to missing data in some columns.

5. Attributes and Methods


Attributes are characteristics or properties that store information about the DataFrame itself and only return stored data from the Dataframe. Therefore, they do not need any parentheses as they simply return data.

Methods, however, are functions that perform operations / calculations associated with a Dataframe. They require parentheses as they execute code.

6. The definitions of each summary statistics of df.describe():

- count produces the number of non-missing (non-NaN) values in each column.
- mean produces the arithmetic mean (average) of the values in each column.
- std (standard Deviation) produces the measurement of the spread or dispersion of the values around the mean.
- min (minimum) produces the smallest value in the column.
- 25% (first quartile) produces the value below which 25% of the data falls. It's the first quartile (Q1), providing information about the lower part of the data distribution.
- 50% (median or second quartile) produces the middle value of the data (50% of the data is below this value). It's the median and is also the second quartile (Q2).
- 75% (third quartile) produces the value below which 75% of the data falls. It's the third quartile (Q3), giving insight into the upper part of the data distribution.
- max (maximum) produces the largest value in the column.

7.1 I am creating a machine learning model that uses the titanic.csv dataset to predict whether passengers survived the Titanic disaster. In this case, the 'age' column has missing values, but it is important as age is a key factor (children given priority for survival). As a result, the 'age' column shouldn't be removed but we also don't want to include rows with missing age values. Therefore, df.dropna() is preferred over del df['col'].

7.2 Working with the same dataset, the 'deck' column has too many missing values to be useful. Even when it's filled, it doesn't provide essential information. Therefore, it is best to just delete the whole column.

7.3 It's important to have del df['col'] first before df.dropna() when used together because it can be used to avoid unnecessary row removal. If I use df.dropna() first, it can also shrink my dataset because it would unintentionally remove rows that could have meaningful information. Overall, having del df['col'] first before df.dropna() would ensure that it preserves data, be efficient, and produce cleaner results.

7.4 First, we need to identify which columns have missing values and assess their importance to the dataset. Next, remove any irrelevant data using del df['col']. Drop rows from important columns using df.dropna(). In this case, we will drop the deck column as it is mostly empty with missing values. Then, we drop rows that have missing important data such as age, embarked, and fare.

Before and after results:

In [1]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

Before Cleanup - Missing Values Per Column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

After Cleanup - Missing Values Per Column:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

Shape of the dataset before cleanup: (891, 14)
Shape of the dataset after cleanup: (712, 14)


In the end, we retrieved as much relevant data as possible. This results in a clean dataset with no missing values.

8.1 Based on the Chatbot session, df.groupby("col1") groups the DataFrame df by the values in col1. This results in a new object where rows are split into groups based on the values of "col1". ["col2"] selects the column col2 from the grouped DataFrame. The analysis that follows will only be applied to this column within each group defined by col1. .describe() generates a summary of descriptive statistics for col2 within each group of col1.

An example given by the Chatbot:

In [7]:
df.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

To demonstrate understanding, another example from the "titanic" dataset is by using the 'parch' and 'fare' columns.

In [6]:
df.groupby("parch")["fare"].describe()

count       mean         std      min       25%       50%      75%  \
parch                                                                       
0      678.0  25.586774   41.878786   0.0000   7.85420  10.50000  26.0000   
1      118.0  46.778180   63.709430   7.2292  17.21250  26.00000  39.6875   
2       80.0  64.337604   65.993088   7.7500  26.00000  32.88125  69.5500   
3        5.0  25.951660    7.862611  18.7500  19.25830  23.00000  34.3750   
4        4.0  84.968750  118.731099  21.0750  26.19375  27.90000  86.6750   
5        5.0  32.550000    4.101134  29.1250  31.27500  31.27500  31.3875   
6        1.0  46.900000         NaN  46.9000  46.90000  46.90000  46.9000   

            max  
parch            
0      512.3292  
1      512.3292  
2      263.0000  
3       34.3750  
4      263.0000  
5       39.6875  
6       46.9000

'parch' is the number of parents and children onboard. 0 is none. 1 is one parent/child and so on. 'fare' is the fare paid by each passenger. 'count' is the number of passengers in that group (e.g., there are 678 passengers with parch = 0). 'mean' is the average fare for that group (e.g., passengers with parch = 0 paid an average fare of $25.586774). 'std' is the standard deviation of the fares for that group, which shows how much the fare values vary. 'min' is the lowest fare paid by someone in that group. '25%' means that 25% of the passengers in that group paid less than this amount. '50%' means that 50% of the passengers in that group paid less than this amount. '75%' means that 75% of passengers in that group paid less than this amount. 'max' is the maximum fare paid by someone in that group.

8.2 df.describe() gives you an overall dataset summary, where 'count' reflects how many values exist per column.
df.groupby("col1")["col2"].describe() is more focused on group-based analysis, with 'count' representing the size of each group. The main goal of the groupby() count is not just to exclude missing data, but to show how the data is distributed among groups, which is essential for making comparisons between categories.

8.3 Assessment of Chatbot vs. Google Performance:

Using code from the "before and after" results from no. 7, we can test out if Chatbot or Google has better results for fixing code errors.

A. Forgetting to include import pandas as pd in my code:

In [1]:
# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

NameError: name 'pd' is not defined

Chatbot response:

In [2]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

Before Cleanup - Missing Values Per Column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

After Cleanup - Missing Values Per Column:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

Shape of the dataset before cleanup: (891, 14)
Shape of the dataset after cleanup: (712, 14)


Result: Fixes code instantly

Google response after searching "NameError: name 'pd' is not defined":

Found solution in: https://stackoverflow.com/questions/40407090/nameerror-name-pd-is-not-defined. Took a bit of reading and some time.

B. Mistype "titanic.csv" as "titanics.csv":

In [1]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

HTTPError: HTTP Error 404: Not Found

Chatbot response:

In [3]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")


Before Cleanup - Missing Values Per Column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

After Cleanup - Missing Values Per Column:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

Shape of the dataset before cleanup: (891, 14)
Shape of the dataset after cleanup: (712, 14)


Result: Found problem and fixes code instantly.

Google response after copy and pasting first paragraph of error code:

Found correct link to dataset: https://github.com/mwaskom/seaborn-data/blob/master/titanic.csv

Other typos in URL yields similar error codes:

In [4]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaburn-data/mastre/titanics.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

HTTPError: HTTP Error 404: Not Found

C. Changing df to DF for every instance in code:

In [5]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = DF.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del DF['deck']

# Step 3: Drop rows with missing values in essential columns
DF_clean = DF.dropna()

# Step 4: Check missing values after cleanup
missing_after = DF_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = DF.shape
shape_after = DF_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

NameError: name 'DF' is not defined

Chatbot response:

In [6]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")


Before Cleanup - Missing Values Per Column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

After Cleanup - Missing Values Per Column:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

Shape of the dataset before cleanup: (891, 14)
Shape of the dataset after cleanup: (712, 14)


Result: changed DF to df in every instance.

Google response after copy and pasting error code:

Couldn't find reasonable answer to my code. Only relevant answer was from: https://pandas.pydata.org/docs/user_guide/missing_data.html, but had to read a lot to find my answer.

D. Code with missing parentheses:

In [7]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum(

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

SyntaxError: invalid syntax. Perhaps you forgot a comma? (4117513225.py, line 5)

Chatbot response:

In [8]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)  # Added missing parenthesis

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()  # Added missing parenthesis

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

Before Cleanup - Missing Values Per Column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

After Cleanup - Missing Values Per Column:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

Shape of the dataset before cleanup: (891, 14)
Shape of the dataset after cleanup: (712, 14)


Result: Fixed both of my missing parentheses, and not only one that is said in the error output.

Google response: Pretty easy to find. All I needed to do was fix missing parentheses. Link: https://stackoverflow.com/questions/74749656/why-am-i-getting-invalid-syntax-perhaps-you-forgot-a-comma

E. Mistype one of the names of the chained functions with the code by changing dropna to dripna:

In [9]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dripna()

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")

AttributeError: 'DataFrame' object has no attribute 'dripna'

Chatbot response:

In [10]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Step 1: Check missing values before cleanup
missing_before = df.isnull().sum()

# Step 2: Remove irrelevant columns with many missing values
del df['deck']

# Step 3: Drop rows with missing values in essential columns
df_clean = df.dropna()  # Fixed typo here

# Step 4: Check missing values after cleanup
missing_after = df_clean.isnull().sum()

# Step 5: Report the shape of the dataset before and after
shape_before = df.shape
shape_after = df_clean.shape

print("Before Cleanup - Missing Values Per Column:")
print(missing_before)
print("\nAfter Cleanup - Missing Values Per Column:")
print(missing_after)

print(f"\nShape of the dataset before cleanup: {shape_before}")
print(f"Shape of the dataset after cleanup: {shape_after}")


Before Cleanup - Missing Values Per Column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

After Cleanup - Missing Values Per Column:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

Shape of the dataset before cleanup: (891, 14)
Shape of the dataset after cleanup: (712, 14)


Result: Changed dripna back to dropna.

Google response after searching "AttributeError":

Found web link that contains information on what I did wrong: https://www.geeksforgeeks.org/python-attributeerror/

F. Using a column name that's not in my data for the groupby and column selection by capitalizing parch -> Parch and fare -> Fare in a separate code:

In [2]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby("Parch")["fare"].describe()

KeyError: 'Parch'

In [3]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby("parch")["Fare"].describe()

KeyError: 'Column not found: Fare'

Chatbot responses:

In [5]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Print the column names to verify
print("Columns in the dataset:")
print(df.columns)

# Check the column names
# If 'Parch' is not listed, then use the correct column name
# For example, if 'parch' is the correct column name, use it in the groupby

df.groupby("parch")["fare"].describe()

Columns in the dataset:
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')


count       mean         std      min       25%       50%      75%  \
parch                                                                       
0      678.0  25.586774   41.878786   0.0000   7.85420  10.50000  26.0000   
1      118.0  46.778180   63.709430   7.2292  17.21250  26.00000  39.6875   
2       80.0  64.337604   65.993088   7.7500  26.00000  32.88125  69.5500   
3        5.0  25.951660    7.862611  18.7500  19.25830  23.00000  34.3750   
4        4.0  84.968750  118.731099  21.0750  26.19375  27.90000  86.6750   
5        5.0  32.550000    4.101134  29.1250  31.27500  31.27500  31.3875   
6        1.0  46.900000         NaN  46.9000  46.90000  46.90000  46.9000   

            max  
parch            
0      512.3292  
1      512.3292  
2      263.0000  
3       34.3750  
4      263.0000  
5       39.6875  
6       46.9000

In [6]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Print column names to verify the correct names
print("Columns in the dataset:")
print(df.columns)

# Use the correct column names for grouping and describing
df_grouped = df.groupby("parch")["fare"].describe()

print(df_grouped)

Columns in the dataset:
Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')
       count       mean         std      min       25%       50%      75%  \
parch                                                                       
0      678.0  25.586774   41.878786   0.0000   7.85420  10.50000  26.0000   
1      118.0  46.778180   63.709430   7.2292  17.21250  26.00000  39.6875   
2       80.0  64.337604   65.993088   7.7500  26.00000  32.88125  69.5500   
3        5.0  25.951660    7.862611  18.7500  19.25830  23.00000  34.3750   
4        4.0  84.968750  118.731099  21.0750  26.19375  27.90000  86.6750   
5        5.0  32.550000    4.101134  29.1250  31.27500  31.27500  31.3875   
6        1.0  46.900000         NaN  46.9000  46.90000  46.90000  46.9000   

            max  
parch            
0      512.3292  
1      512.3292  
2      263.0000  
3  

Results: Both errors got returned with correct code fixes.

Google responses after searching about both errors:

Searching "KeyError" yields relevant web link: https://stackoverflow.com/questions/10116518/im-getting-key-error-in-python

Searching "KeyError: 'Column not found: Fare'" yields relevant web link: https://stackoverflow.com/questions/61987268/key-error-column-not-found-but-its-there

G. Forgetting to put string in quotes to column names for the groupby and column selection:

In [7]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby(parch)["fare"].describe()

NameError: name 'parch' is not defined

In [8]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby("parch")[fare].describe()

NameError: name 'fare' is not defined

Chatbot responses:

In [9]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Corrected code with 'parch' in quotes
df_grouped = df.groupby("parch")["fare"].describe()

print(df_grouped)

       count       mean         std      min       25%       50%      75%  \
parch                                                                       
0      678.0  25.586774   41.878786   0.0000   7.85420  10.50000  26.0000   
1      118.0  46.778180   63.709430   7.2292  17.21250  26.00000  39.6875   
2       80.0  64.337604   65.993088   7.7500  26.00000  32.88125  69.5500   
3        5.0  25.951660    7.862611  18.7500  19.25830  23.00000  34.3750   
4        4.0  84.968750  118.731099  21.0750  26.19375  27.90000  86.6750   
5        5.0  32.550000    4.101134  29.1250  31.27500  31.27500  31.3875   
6        1.0  46.900000         NaN  46.9000  46.90000  46.90000  46.9000   

            max  
parch            
0      512.3292  
1      512.3292  
2      263.0000  
3       34.3750  
4      263.0000  
5       39.6875  
6       46.9000  


In [10]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Corrected code with 'fare' in quotes
df_grouped = df.groupby("parch")["fare"].describe()

print(df_grouped)

       count       mean         std      min       25%       50%      75%  \
parch                                                                       
0      678.0  25.586774   41.878786   0.0000   7.85420  10.50000  26.0000   
1      118.0  46.778180   63.709430   7.2292  17.21250  26.00000  39.6875   
2       80.0  64.337604   65.993088   7.7500  26.00000  32.88125  69.5500   
3        5.0  25.951660    7.862611  18.7500  19.25830  23.00000  34.3750   
4        4.0  84.968750  118.731099  21.0750  26.19375  27.90000  86.6750   
5        5.0  32.550000    4.101134  29.1250  31.27500  31.27500  31.3875   
6        1.0  46.900000         NaN  46.9000  46.90000  46.90000  46.9000   

            max  
parch            
0      512.3292  
1      512.3292  
2      263.0000  
3       34.3750  
4      263.0000  
5       39.6875  
6       46.9000  


Results: Both errors got returned with correct code fixes.

Google responses after searching about both errors:

Searching "Python NameError" yields relevant web link: https://stackoverflow.com/questions/14804084/python-nameerror-name-is-not-defined

Summary: Based on the assessment experiment, ChatGPT can correct code almost instantly while Googling takes a lot of time to find the solution to the code errors.

9. Yes, I reviewed the course wiki-textbook and interacted with a Chatbot for this homework.

Chatbot Log Histories:

A. Error code exercise chat summary:

link: https://chatgpt.com/share/66e386ad-3800-8004-a9c0-e6ccb8ae34fe

Certainly! Here’s a summary of our session:

1. **Code Issues**:
   - We explored various errors in your code related to loading and cleaning the Titanic dataset. These included `NameError` and `KeyError` due to incorrect column names and syntax issues.
   - You encountered errors like `HTTPError` when trying to access data from an incorrect URL and `KeyError` when column names used in operations were not found in the DataFrame.

2. **Correcting Errors**:
   - We corrected issues related to invalid column names and syntax errors. For instance, ensuring column names like `'fare'` are used with correct capitalization and enclosing column names in quotes.

3. **Best Practices**:
   - Always check the actual column names in the DataFrame using `df.columns` to avoid referencing non-existent columns.
   - Use quotes around column names when indexing or performing operations on DataFrame columns.

If you have any more questions or need further assistance, feel free to ask!

B. No. 8 Chatbot Session:

link: https://chatgpt.com/share/66e3867f-dea4-8004-8dd6-ec05e4f30031

Here’s a summary of our session:

1. **Titanic Dataset Analysis**:
   - **`df.groupby("col1")["col2"].describe()`**: Groups the DataFrame by `col1` and generates descriptive statistics for `col2` within each group. The statistics include count, mean, standard deviation, min, quartiles, and max.
   - **`df.groupby("parch")["fare"].describe()`**: Groups the Titanic dataset by `parch` (number of parents/children aboard) and provides descriptive statistics for the `fare` column for each group. This includes count, mean, standard deviation, min, 25th percentile, median, 75th percentile, and max of fares for each group defined by `parch`.

2. **Understanding Output**:
   - Each group in `parch` will have its own set of statistics summarizing the fare distribution. This helps analyze how the fare varies with the number of parents/children aboard.

If you have any more questions or need further clarification, feel free to ask!

C. Overall HW1 Chatbot Session:

link: https://chatgpt.com/share/66e3910a-3808-8004-bf48-382cc4845e74

In this session, we covered several key concepts related to data analysis with pandas, particularly using the Titanic dataset. Here's a summary:

1. **Counting Rows and Columns in a DataFrame**: We explored how to check the shape of a dataset using `df.shape` to get the number of rows and columns.

2. **Understanding Observations and Variables**: We discussed the general definitions of observations (rows) and variables (columns) in a dataset.

3. **Summarizing Data**: You learned how to generate summary statistics with `df.describe()` and count unique values with `df['column'].value_counts()`.

4. **Handling Missing Data**: We examined the presence of missing values in the Titanic dataset and discussed methods like `df.dropna()` to remove missing data, as well as `del df['col']` to delete entire columns that are irrelevant or contain too many missing values.

5. **Comparison of `df.describe()` and `groupby()`**: We explored the differences between the `count` values in `df.describe()` and those in `df.groupby("col1")["col2"].describe()`. The focus was on how `df.describe()` handles missing values globally while `groupby()` focuses on group-wise analysis.

6. **Efficient Data Cleaning**: We used a combination of `del df['col']` and `df.dropna()` to efficiently clean the dataset, explaining when one method would be preferred over the other.

7. **Use Cases**: We discussed scenarios where removing irrelevant columns before handling missing values is important to avoid unnecessary data loss and improve efficiency.

Would you like to continue or revisit any specific part of our discussion?

End of HW1.